CreeDictionary/shared/__init__.py

```diff
@@ -1,30 +1,153 @@
+
+    def analyze(self, cree: str) -> Set[Analysis]:
+        return feed_to_hfstol_process(self.proc, cree)
+
+    def __new__(cls, **kwargs):
+        if cls._instance is None:
+            _instance = super().__new__(cls, **kwargs)
+        else:
+            _instance = cls._instance
+        return _instance
+
+
+# Pythonic singleton. The initialization creates a demonic hfstol process waiting to be used at any time
+class NormativeGenerator:
+    _instance = None
+
+    def __init__(self):
+        if NormativeGenerator._instance is None:
+            # do initialization here
+            self.proc = subprocess.Popen(
+                args=[
+                    "hfst-optimized-lookup",
+                    "--quiet",
+                    "--pipe-mode",
+                    str(shared_res_dir / "fst" / "crk-normative-generator.hfstol"),
+                ],
+                stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                shell=False,
+                universal_newlines=True,
+            )
+            NormativeGenerator._instance = self
+
+    def generate(self, analysis: Analysis) -> Set[str]:
+        return feed_to_hfstol_process(self.proc, analysis)
+
+    def __new__(cls, **kwargs):
+        if cls._instance is None:
+            _instance = super().__new__(cls, **kwargs)
+        else:
+            _instance = cls._instance
+        return _instance
+
+
+## helper function
+def feed_to_hfstol_process(proc: subprocess.Popen, surface_form) -> set:
+
+    # example output of the hfstol binary
+
+    # $ hfst-optimized-lookup --quiet --pipe-mode CreeDictionary/res/fst/crk-descriptive-analyzer.hfstol
+    # nipa (this is input line)
+    # nipa\tnipâw+V+AI+Imp+Imm+2Sg
+    #
+    # niskak (this is input line)
+    # niskak\tniska+N+A+Pl
+    # niskak\tnîskâw+V+II+Cnj+Prs+3Sg
+    #
+    # c
+    # c       c       +?
+    #
+    #
+    #
+    # EOF
+
+    # two or more consecutive newlines signals the end of the output
+    assert proc.stdin is not None
+    proc.stdin.write(surface_form + "\n")
+    proc.stdin.flush()
+
+    results: Set[str] = set()
+
+    consecutive_newline_count = 0
+
+    assert proc.stdout is not None
+    # ignore empty characters (consecutive white space characters from previous invocations)
+    while 1:
+        char = proc.stdout.read(1)
+        if char:
+            line_buffer = char
+            break
+
+    while consecutive_newline_count != 2:
+        new_char = proc.stdout.read(1)
+
+        if new_char == "\n":
+            consecutive_newline_count += 1
+            if line_buffer.strip():
+                print("lol", line_buffer)
+                original_input, res, *rest = line_buffer.split("\t")
+                if "+?" not in rest and "+?" not in res:
+                    results.add(res)
+            line_buffer = ""
+        else:
+            line_buffer += new_char
+            consecutive_newline_count = 0
```

Comment:
I believe I screwed up this output parsing part in `hfstol` pacakge, so that with `hfstol` package, demon processes' stdout has leftover output from previous inputs. Anyways, I believe I nailed the correct version is achieved here!

I'd like to also add optional multi-processing here later so that we can scrap `hfstol` package altogether. A standalone package feels kinda hard to maintain and debug. Everytime I changed something in the `hfstol` repo, I have to publish a new release and create a pull request here to test it in CI, which is tremendously inefficient.

Reply:
> Anyways, I believe I nailed it here.

@eddieantonio You are welcome to re-run integration tests multiple times to confirm :+1:

Reply:
that ain't it, chief

Reply:
@eddieantonio so it's still flaky..wtf, it's rerun twice, the "merge" integration test consistently fails , but not the "branch" integration. (or maybe it's by pure chance)

I have no idea what's happening. :crying_cat_face: 

Could you help and have a look?